File: compilationTests/src/test/java/androidx/databinding/compilationTest/SimpleCompilationTest.java
/*
 * Copyright (C) 2015 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.databinding.compilationTest;

import android.databinding.tool.processing.ErrorMessages;
import android.databinding.tool.processing.ScopedErrorReport;
import android.databinding.tool.processing.ScopedException;
import android.databinding.tool.store.Location;
import org.hamcrest.CoreMatchers;
import org.hamcrest.MatcherAssert;
import org.junit.Ignore;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import java.io.File;
import java.io.IOException;
import java.net.URISyntaxException;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import static androidx.databinding.compilationTest.DataBindingCompilationTestUtilsKt.copyResourceWithReplacement;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

@SuppressWarnings("ThrowableResultOfMethodCallIgnored")
@RunWith(JUnit4.class)
public class SimpleCompilationTest extends BaseCompilationTest {

    @Test
    public void testMultipleExceptionsInDifferentFiles()
            throws IOException, URISyntaxException, InterruptedException {
        prepareProject();
        copyResourceTo("/layout/undefined_variable_binding.xml",
                "/app/src/main/res/layout/broken.xml");
        copyResourceTo("/layout/invalid_setter_binding.xml",
                "/app/src/main/res/layout/invalid_setter.xml");
        CompilationResult result = runGradle("assembleDebug");
        assertNotEquals(result.output, 0, result.resultCode);
        List<ScopedException> bindingExceptions = result.getBindingExceptions();
        assertEquals(result.error, 2, bindingExceptions.size());
        File broken = new File(testFolder, "/app/src/main/res/layout/broken.xml");
        File invalidSetter = new File(testFolder, "/app/src/main/res/layout/invalid_setter.xml");
        for (ScopedException exception : bindingExceptions) {
            ScopedErrorReport report = exception.getScopedErrorReport();
            final File errorFile = requireErrorFile(report);
            String message = null;
            String expectedErrorFile = null;
            if (errorFile.getCanonicalPath().equals(broken.getCanonicalPath())) {
                message = String.format(ErrorMessages.UNDEFINED_VARIABLE, "myVariable");
                expectedErrorFile = "/app/src/main/res/layout/broken.xml";
            } else if (errorFile.getCanonicalPath().equals(invalidSetter.getCanonicalPath())) {
                message = String.format(ErrorMessages.CANNOT_FIND_SETTER_CALL,
                    "android.widget.TextView", "android:textx", String.class.getCanonicalName());
                expectedErrorFile = "/app/src/main/res/layout/invalid_setter.xml";
            } else {
                fail("unexpected exception " + exception.getBareMessage());
            }
            assertEquals(1, report.getLocations().size());
            Location loc = report.getLocations().get(0);
            String extract = extract(expectedErrorFile, loc);
            assertEquals("myVariable", extract);
            assertEquals(message, exception.getBareMessage());
        }
    }

    @Test
    public void testSingleModule() throws IOException, URISyntaxException, InterruptedException {
        prepareApp(toMap(KEY_DEPENDENCIES, "implementation project(':module1')",
                KEY_SETTINGS_INCLUDES, "include ':app'\ninclude ':module1'"));
        prepareModule("module1", "com.example.module1", toMap());
        copyResourceTo("/layout/basic_layout.xml", "/module1/src/main/res/layout/module_layout.xml");
        copyResourceTo("/layout/basic_layout.xml", "/app/src/main/res/layout/app_layout.xml");
        CompilationResult result = runGradle("assembleDebug");
        assertEquals(result.error, 0, result.resultCode);
    }

    @Test
    public void testConflictingIds() throws IOException, URISyntaxException, InterruptedException {
        prepareProject();
        copyResourceTo("/layout/duplicate_ids.xml",
                "/app/src/main/res/layout/duplicate_ids.xml");
        CompilationResult result = runGradle("assembleDebug");
        List<ScopedException> exceptions = result.getBindingExceptions();
        assertMessages(exceptions,
                String.format(ErrorMessages.DUPLICATE_VIEW_OR_INCLUDE_ID,
                        "TextView", "@+id/shared_id"),
                String.format(ErrorMessages.DUPLICATE_VIEW_OR_INCLUDE_ID,
                        "TextView", "@+id/shared_id"));
    }

    @Test
    public void testConflictingIds_include() throws IOException, URISyntaxException, InterruptedException {
        prepareProject();
        copyResourceTo("/layout/basic_layout.xml", "/app/src/main/res/layout/basic_layout.xml");
        copyResourceTo("/layout/duplicate_include_ids.xml",
                "/app/src/main/res/layout/duplicate_include_ids.xml");
        CompilationResult result = runGradle("assembleDebug");
        List<ScopedException> exceptions = result.getBindingExceptions();
        assertMessages(exceptions,
                String.format(ErrorMessages.DUPLICATE_VIEW_OR_INCLUDE_ID,
                        "include", "@+id/shared_id"),
                String.format(ErrorMessages.DUPLICATE_VIEW_OR_INCLUDE_ID,
                        "include", "@+id/shared_id"));
    }

    @Test
    public void testConflictingIds_includeVsView() throws IOException, URISyntaxException, InterruptedException {
        prepareProject();
        copyResourceTo("/layout/basic_layout.xml", "/app/src/main/res/layout/basic_layout.xml");
        copyResourceTo("/layout/duplicate_include_vs_view_ids.xml",
                "/app/src/main/res/layout/duplicate_include_vs_view_ids.xml");
        CompilationResult result = runGradle("assembleDebug");
        List<ScopedException> exceptions = result.getBindingExceptions();
        assertMessages(exceptions,
                String.format(ErrorMessages.DUPLICATE_VIEW_OR_INCLUDE_ID,
                        "TextView", "@+id/shared_id"),
                String.format(ErrorMessages.DUPLICATE_VIEW_OR_INCLUDE_ID,
                        "include", "@+id/shared_id"));
    }

    private void assertMessages(List<ScopedException> exceptions,
                                String... messages) {
        List<String> actual = exceptions.stream()
                .map(ScopedException::getBareMessage)
                .collect(Collectors.toList());
        MatcherAssert.assertThat(actual, CoreMatchers.hasItems(messages));
        MatcherAssert.assertThat(actual.size(), CoreMatchers.is(messages.length));
    }

    // TODO: reenable this test once it works.
    @Ignore
    public void testModuleDependencyChange() throws IOException, URISyntaxException,
            InterruptedException {
        prepareApp(toMap(KEY_DEPENDENCIES, "implementation project(':module1')",
                KEY_SETTINGS_INCLUDES, "include ':app'\ninclude ':module1'"));
        prepareModule("module1", "com.example.module1", toMap(
                KEY_DEPENDENCIES, "implementation 'com.android.support:appcompat-v7:23.1.1'"
        ));
        copyResourceTo("/layout/basic_layout.xml", "/module1/src/main/res/layout/module_layout.xml");
        copyResourceTo("/layout/basic_layout.xml", "/app/src/main/res/layout/app_layout.xml");
        CompilationResult result = runGradle("assembleDebug");
        assertEquals(result.error, 0, result.resultCode);
        File moduleFolder = new File(testFolder, "module1");
        copyResourceWithReplacement("/module_build.gradle", new File(moduleFolder, "build.gradle"),
                toMap());
        result = runGradle("assembleDebug");
        assertEquals(result.error, 0, result.resultCode);
    }

    @Test
    public void testTwoLevelDependency() throws IOException, URISyntaxException, InterruptedException {
        prepareApp(toMap(KEY_DEPENDENCIES, "implementation project(':module1')",
                KEY_SETTINGS_INCLUDES, "include ':app'\ninclude ':module1'\n"
                        + "include ':module2'"));
        prepareModule("module1", "com.example.module1", toMap(KEY_DEPENDENCIES,
                "implementation project(':module2')"));
        prepareModule("module2", "com.example.module2", toMap());
        copyResourceTo("/layout/basic_layout.xml",
                "/module2/src/main/res/layout/module2_layout.xml");
        copyResourceTo("/layout/basic_layout.xml", "/module1/src/main/res/layout/module1_layout.xml");
        copyResourceTo("/layout/basic_layout.xml", "/app/src/main/res/layout/app_layout.xml");
        CompilationResult result = runGradle("assembleDebug");
        assertEquals(result.error, 0, result.resultCode);
    }

    @Test
    public void testIncludeInMerge() throws Throwable {
        prepareProject();
        copyResourceTo("/layout/merge_include.xml", "/app/src/main/res/layout/merge_include.xml");
        CompilationResult result = runGradle("assembleDebug");
        assertNotEquals(0, result.resultCode);
        List<ScopedException> errors = ScopedException.extractErrors(result.error);
        assertEquals(result.error, 1, errors.size());
        final ScopedException ex = errors.get(0);
        final ScopedErrorReport report = ex.getScopedErrorReport();
        final File errorFile = requireErrorFile(report);
        assertEquals(
                new File(testFolder, "/app/src/main/res/layout/merge_include.xml")
                        .getCanonicalFile(),
                errorFile.getCanonicalFile());
        assertEquals(ErrorMessages.INCLUDE_INSIDE_MERGE, ex.getBareMessage());
    }

    @Test
    public void testAssignTwoWayEvent() throws Throwable {
        prepareProject();
        copyResourceTo("/layout/layout_with_two_way_event_attribute.xml",
                "/app/src/main/res/layout/layout_with_two_way_event_attribute.xml");
        CompilationResult result = runGradle("assembleDebug");
        assertNotEquals(0, result.resultCode);
        List<ScopedException> errors = ScopedException.extractErrors(result.error);
        assertEquals(result.error, 1, errors.size());
        final ScopedException ex = errors.get(0);
        final ScopedErrorReport report = ex.getScopedErrorReport();
        final File errorFile = requireErrorFile(report);
        assertEquals(new File(testFolder,
                "/app/src/main/res/layout/layout_with_two_way_event_attribute.xml")
                        .getCanonicalFile(),
                errorFile.getCanonicalFile());
        assertEquals(
            String.format(ErrorMessages.TWO_WAY_EVENT_ATTRIBUTE, "android:textAttrChanged"),
            ex.getBareMessage());
    }

    @Test
    public void testDependantDoesNotExist() throws Throwable {
        prepareProject();
        copyResourceTo("/layout/layout_with_dependency.xml",
                "/app/src/main/res/layout/layout_with_dependency.xml");
        copyResourceTo(
                "/androidx/databinding/compilationTest/badJava/ObservableNoDependent.java",
                "/app/src/main/java/androidx/databinding/compilationTest/badJava/MyObservable.java");

        CompilationResult result = runGradle("assembleDebug");
        assertNotEquals(0, result.resultCode);
        List<ScopedException> errors = ScopedException.extractErrors(result.error);
        assertEquals(result.error, 1, errors.size());
        final ScopedException ex = errors.get(0);
        final ScopedErrorReport report = ex.getScopedErrorReport();
        final File errorFile = requireErrorFile(report);
        assertEquals(new File(testFolder,
                        "/app/src/main/res/layout/layout_with_dependency.xml")
                        .getCanonicalFile(),
                errorFile.getCanonicalFile());
        assertEquals("Could not find dependent property 'notExist' referenced in " +
                        "@Bindable annotation on " +
                        "androidx.databinding.compilationTest.badJava.MyObservable.getField",
                ex.getBareMessage());
    }

    @Test
    public void testDependantNotBindable() throws Throwable {
        prepareProject();
        copyResourceTo("/layout/layout_with_dependency.xml",
                "/app/src/main/res/layout/layout_with_dependency.xml");
        copyResourceTo(
                "/androidx/databinding/compilationTest/badJava/ObservableNotBindableDependent.java",
                "/app/src/main/java/androidx/databinding/compilationTest/badJava/MyObservable.java");

        CompilationResult result = runGradle("assembleDebug");
        assertNotEquals(0, result.resultCode);
        List<ScopedException> errors = ScopedException.extractErrors(result.error);
        assertEquals(result.error, 1, errors.size());
        final ScopedException ex = errors.get(0);
        final ScopedErrorReport report = ex.getScopedErrorReport();
        final File errorFile = requireErrorFile(report);
        assertEquals(new File(testFolder,
                        "/app/src/main/res/layout/layout_with_dependency.xml")
                        .getCanonicalFile(),
                errorFile.getCanonicalFile());
        assertEquals("The dependent property 'otherField' referenced in " +
                "@Bindable annotation on " +
                "androidx.databinding.compilationTest.badJava.MyObservable.getField " +
                "must be annotated with @Bindable", ex.getBareMessage());
    }

    @Test
    public void testDependantField() throws Throwable {
        prepareProject();
        copyResourceTo("/layout/layout_with_dependency.xml",
                "/app/src/main/res/layout/layout_with_dependency.xml");
        copyResourceTo(
                "/androidx/databinding/compilationTest/badJava/ObservableFieldDependent.java",
                "/app/src/main/java/androidx/databinding/compilationTest/badJava/MyObservable.java");

        CompilationResult result = runGradle("assembleDebug");
        assertNotEquals(0, result.resultCode);
        List<ScopedException> errors = ScopedException.extractErrors(result.error);
        assertEquals(result.error, 1, errors.size());
        final ScopedException ex = errors.get(0);
        final ScopedErrorReport report = ex.getScopedErrorReport();
        final File errorFile = requireErrorFile(report);
        assertTrue(errorFile.exists());
        assertEquals(new File(testFolder,
                        "/app/src/main/res/layout/layout_with_dependency.xml")
                        .getCanonicalFile(),
                errorFile.getCanonicalFile());
        assertEquals("Bindable annotation with property names is only supported on methods. " +
                "Field 'androidx.databinding.compilationTest.badJava.MyObservable.field' has " +
                "@Bindable(\"otherField\")", ex.getBareMessage());
    }
}

File: compilationTests/src/test/java/androidx/databinding/compilationTest/bazel/SimpleCompilationTest.kt
/*
 * Copyright (C) 2021 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package androidx.databinding.compilationTest.bazel

import android.databinding.tool.processing.ErrorMessages
import android.databinding.tool.processing.ScopedErrorReport
import android.databinding.tool.processing.ScopedException
import android.databinding.tool.store.Location
import androidx.databinding.compilationTest.BaseCompilationTest
import androidx.databinding.compilationTest.BaseCompilationTest.KEY_DEPENDENCIES
import androidx.databinding.compilationTest.BaseCompilationTest.KEY_MANIFEST_PACKAGE
import androidx.databinding.compilationTest.CompilationResult
import com.google.common.base.Joiner
import org.apache.commons.io.FileUtils
import org.apache.commons.io.filefilter.NameFileFilter
import org.apache.commons.io.filefilter.PrefixFileFilter
import org.hamcrest.CoreMatchers
import org.hamcrest.MatcherAssert
import org.junit.Assert
import org.junit.Ignore
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import java.io.File
import java.io.IOException
import java.nio.charset.StandardCharsets

@RunWith(JUnit4::class)
class SimpleCompilationTest : DataBindingCompilationTestCase() {

    @Test
    fun listTasks() {
        loadApp()
        val result = invokeGradleTasks(project, "tasks")
        assertTrue("Empty project tasks failed", result.isBuildSuccessful)
    }

    @Test
    fun testEmptyCompilation() {
        loadApp()
        val result = invokeGradleTasks(project, "assembleDebug")
        assertTrue("Basic compile failed", result.isBuildSuccessful)
    }

    @Test
    fun testMultipleConfigs() {
        loadApp()
        copyTestData(
            "layout/basic_layout.xml",
            "app/src/main/res/layout/main.xml"
        )
        copyTestData(
            "layout/basic_layout.xml",
            "app/src/main/res/layout-sw100dp/main.xml"
        )
        val result = invokeGradleTasks(project, "assembleDebug")
        assertTrue(result.isBuildSuccessful)
        val debugOut = File(
            projectRoot,
            "app/build/intermediates/incremental/mergeDebugResources/stripped.dir"
        )
        val layoutFiles = FileUtils.listFiles(
            debugOut, NameFileFilter("main.xml"),
            PrefixFileFilter("layout")
        )
        assertTrue("Unexpected generated layout count", layoutFiles.size > 1)
        for (layout: File in layoutFiles) {
            val contents = FileUtils.readFileToString(layout, StandardCharsets.UTF_8)
            if (layout.parent.contains("sw100")) {
                assertTrue(
                    "File has wrong tag:" + layout.path,
                    contents.indexOf("android:tag=\"layout-sw100dp/main_0\"") > 0
                )
            } else {
                assertTrue(
                    "File has wrong tag:" + layout.path + "\n" + contents,
                    contents.indexOf("android:tag=\"layout/main_0\"") > 0
                )
            }
        }
    }

    @Test
    fun testBadSyntax() {
        singleFileErrorTest(
            "layout/layout_with_bad_syntax.xml",
            "app/src/main/res/layout/broken.xml",
            "myVar.length())",
            String.format(
                ErrorMessages.SYNTAX_ERROR,
                "extraneous input ')' expecting {<EOF>, ',', '.', '::', '[', '+', '-', " +
                        "'*', '/', '%', '<<', '>>>', '>>', '<=', '>=', '>', '<', " +
                        "'instanceof', '==', '!=', '&', '^', '|', '&&', '||', '?', '??'}"
            )
        )
    }

    @Test
    fun testBrokenSyntax() {
        singleFileErrorTest(
            "layout/layout_with_completely_broken_syntax.xml",
            "app/src/main/res/layout/broken.xml",
            "new String()", String.format(
                ErrorMessages.SYNTAX_ERROR,
                "mismatched input 'String' expecting {<EOF>, ',', '.', '::', '[', '+', " +
                        "'-', '*', '/', '%', '<<', '>>>', '>>', '<=', '>=', '>', '<', " +
                        "'instanceof', '==', '!=', '&', '^', '|', '&&', '||', '?', '??'}"
            )
        )
    }

    @Test
    fun testUndefinedVariable() {
        singleFileErrorTest(
            "layout/undefined_variable_binding.xml",
            "app/src/main/res/layout/broken.xml", "myVariable", String.format(
                ErrorMessages.UNDEFINED_VARIABLE, "myVariable"
            )
        )
    }

    @Test
    fun testInvalidSetterBinding() {
        singleFileErrorTest(
            "layout/invalid_setter_binding.xml",
            "app/src/main/res/layout/invalid_setter.xml", "myVariable", String.format(
                ErrorMessages.CANNOT_FIND_SETTER_CALL, "android.widget.TextView",
                "android:textx",
                String::class.java.canonicalName
            )
        )
    }

    @Test
    fun testCallbackArgumentCountMismatch() {
        singleFileErrorTest(
            "layout/layout_with_missing_callback_args.xml",
            "app/src/main/res/layout/broken.xml",
            "(seekBar, progress) -> obj.length()", String.format(
                ErrorMessages.CALLBACK_ARGUMENT_COUNT_MISMATCH,
                "androidx.databinding.adapters.SeekBarBindingAdapter.OnProgressChanged",
                "onProgressChanged", 3, 2
            )
        )
    }

    @Test
    fun testDuplicateCallbackArgument() {
        singleFileErrorTest(
            "layout/layout_with_duplicate_callback_identifier.xml",
            "app/src/main/res/layout/broken.xml",
            "(seekBar, progress, progress) -> obj.length()", String.format(
                ErrorMessages.DUPLICATE_CALLBACK_ARGUMENT,
                "progress"
            )
        )
    }

    @Test
    fun testInvalidVariableType() {
        singleFileErrorTest(
            "layout/invalid_variable_type.xml",
            "app/src/main/res/layout/invalid_variable.xml",
            "myVariable",
            String.format(ErrorMessages.CANNOT_RESOLVE_TYPE, "myVariable")
        )
    }

    @Test
    fun testConflictWithVariableName() {
        singleFileWarningTest(
            "layout/layout_with_same_name_for_var_and_callback.xml",
            "app/src/main/res/layout/broken.xml", String.format(
                ErrorMessages.CALLBACK_VARIABLE_NAME_CLASH,
                "myVar", "String", "myVar"
            )
        )
    }

    @Test
    fun testMultipleExceptionsInDifferentFiles() {
        loadApp()
        copyTestData(
            "layout/undefined_variable_binding.xml",
            "app/src/main/res/layout/broken.xml"
        )
        copyTestData(
            "layout/invalid_setter_binding.xml",
            "app/src/main/res/layout/invalid_setter.xml"
        )
        val result = assembleDebug()
        Assert.assertNotEquals(result.output, 0, result.resultCode.toLong())
        val bindingExceptions = result.bindingExceptions
        Assert.assertEquals(result.error, 2, bindingExceptions.size.toLong())
        val broken = File(projectRoot, "/app/src/main/res/layout/broken.xml")
        val invalidSetter = File(projectRoot, "/app/src/main/res/layout/invalid_setter.xml")
        for (exception in bindingExceptions) {
            val report = exception.scopedErrorReport
            val errorFile = requireErrorFile(report)
            var message: String? = null
            var expectedErrorFile: String? = null
            when (errorFile.canonicalPath) {
                broken.canonicalPath -> {
                    message = String.format(ErrorMessages.UNDEFINED_VARIABLE, "myVariable")
                    expectedErrorFile = "/app/src/main/res/layout/broken.xml"
                }
                invalidSetter.canonicalPath -> {
                    message = String.format(
                        ErrorMessages.CANNOT_FIND_SETTER_CALL,
                        "android.widget.TextView", "android:textx", String::class.java.canonicalName
                    )
                    expectedErrorFile = "/app/src/main/res/layout/invalid_setter.xml"
                }
                else -> {
                    Assert.fail("unexpected exception " + exception.bareMessage)
                }
            }
            Assert.assertEquals(1, report.locations.size.toLong())
            val loc = report.locations[0]
            val extract = extract(expectedErrorFile!!, loc)
            Assert.assertEquals("myVariable", extract)
            Assert.assertEquals(message, exception.bareMessage)
        }
    }

    @Test
    fun testSingleModule() {
        loadApp(
            mapOf(
                KEY_DEPENDENCIES to "implementation project(':module1')",
                BaseCompilationTest.KEY_SETTINGS_INCLUDES to "include ':app'\ninclude ':module1'"
            )
        )
        loadModule("module1", mapOf(KEY_MANIFEST_PACKAGE to "com.example.module1"))
        copyTestData("layout/basic_layout.xml", "module1/src/main/res/layout/module_layout.xml")
        copyTestData("layout/basic_layout.xml", "app/src/main/res/layout/app_layout.xml")

        val result: CompilationResult = assembleDebug()
        assertEquals(result.error, 0, result.resultCode.toLong())
    }

    @Test
    fun testConflictingIds() {
        loadApp()
        copyTestData(
            "layout/duplicate_ids.xml",
            "app/src/main/res/layout/duplicate_ids.xml"
        )
        val result = assembleDebug()
        val exceptions = result.bindingExceptions
        assertMessages(
            exceptions, String.format(
                ErrorMessages.DUPLICATE_VIEW_OR_INCLUDE_ID,
                "TextView", "@+id/shared_id"
            ), String.format(
                ErrorMessages.DUPLICATE_VIEW_OR_INCLUDE_ID,
                "TextView", "@+id/shared_id"
            )
        )
    }

    @Test
    fun testConflictingIds_include() {
        loadApp()
        copyTestData("layout/basic_layout.xml", "app/src/main/res/layout/basic_layout.xml")
        copyTestData(
            "layout/duplicate_include_ids.xml",
            "app/src/main/res/layout/duplicate_include_ids.xml"
        )
        val result = assembleDebug()
        val exceptions = result.bindingExceptions
        assertMessages(
            exceptions, String.format(
                ErrorMessages.DUPLICATE_VIEW_OR_INCLUDE_ID,
                "include", "@+id/shared_id"
            ), String.format(
                ErrorMessages.DUPLICATE_VIEW_OR_INCLUDE_ID,
                "include", "@+id/shared_id"
            )
        )
    }

    @Test
    fun testConflictingIds_includeVsView() {
        loadApp()
        copyTestData("layout/basic_layout.xml", "app/src/main/res/layout/basic_layout.xml")
        copyTestData(
            "layout/duplicate_include_vs_view_ids.xml",
            "app/src/main/res/layout/duplicate_include_vs_view_ids.xml"
        )
        val result = assembleDebug()
        val exceptions = result.bindingExceptions
        assertMessages(
            exceptions, String.format(
                ErrorMessages.DUPLICATE_VIEW_OR_INCLUDE_ID,
                "TextView", "@+id/shared_id"
            ), String.format(
                ErrorMessages.DUPLICATE_VIEW_OR_INCLUDE_ID,
                "include", "@+id/shared_id"
            )
        )
    }

    // TODO: reenable this test once it works.
    @Ignore
    fun testModuleDependencyChange() {
        loadApp(
            mapOf(
                KEY_DEPENDENCIES to "implementation project(':module1')",
                BaseCompilationTest.KEY_SETTINGS_INCLUDES to "include ':app'\ninclude ':module1'"
            )
        )
        loadModule(
            "module1", mapOf(
                KEY_DEPENDENCIES to "implementation 'com.android.support:appcompat-v7:23.1.1'",
                KEY_MANIFEST_PACKAGE to "com.example.module1"
            )
        )
        copyTestData("layout/basic_layout.xml", "module1/src/main/res/layout/module_layout.xml")
        copyTestData("layout/basic_layout.xml", "app/src/main/res/layout/app_layout.xml")
        var result = assembleDebug()
        Assert.assertEquals(result.error, 0, result.resultCode.toLong())
        copyTestDataWithReplacement("module_build.gradle", "module1/build.gradle")
        result = assembleDebug()
        Assert.assertEquals(result.error, 0, result.resultCode.toLong())
    }

    @Test
    fun testTwoLevelDependency() {
        loadApp(
            mapOf(
                KEY_DEPENDENCIES to "implementation project(':module1')",
                BaseCompilationTest.KEY_SETTINGS_INCLUDES to
                        "include ':app'\ninclude ':module1'\n"
                        + "include ':module2'"
            )
        )
        loadModule(
            "module1",
            mapOf(
                KEY_DEPENDENCIES to "implementation project(':module2')",
                KEY_MANIFEST_PACKAGE to "com.example.module1"
            )
        )
        loadModule("module2", mapOf(KEY_MANIFEST_PACKAGE to "com.example.module2"))
        copyTestData(
            "layout/basic_layout.xml",
            "module2/src/main/res/layout/module2_layout.xml"
        )
        copyTestData(
            "layout/basic_layout.xml",
            "module1/src/main/res/layout/module1_layout.xml"
        )
        copyTestData("layout/basic_layout.xml", "app/src/main/res/layout/app_layout.xml")
        val result = assembleDebug()
        Assert.assertEquals(result.error, 0, result.resultCode.toLong())
    }

    @Test
    fun testIncludeInMerge() {
        loadApp()
        copyTestData("layout/merge_include.xml", "app/src/main/res/layout/merge_include.xml")
        val result = assembleDebug()
        Assert.assertNotEquals(0, result.resultCode.toLong())
        val errors = ScopedException.extractErrors(result.error)
        Assert.assertEquals(result.error, 1, errors.size.toLong())
        val ex = errors[0]
        val report = ex.scopedErrorReport
        val errorFile = requireErrorFile(report)
        Assert.assertEquals(
            File(projectRoot, "/app/src/main/res/layout/merge_include.xml").canonicalFile,
            errorFile.canonicalFile
        )
        Assert.assertEquals(ErrorMessages.INCLUDE_INSIDE_MERGE, ex.bareMessage)
    }

    @Test
    fun testAssignTwoWayEvent() {
        loadApp()
        copyTestData(
            "layout/layout_with_two_way_event_attribute.xml",
            "app/src/main/res/layout/layout_with_two_way_event_attribute.xml"
        )
        val result: CompilationResult = assembleDebug()
        Assert.assertNotEquals(0, result.resultCode.toLong())
        val errors = ScopedException.extractErrors(result.error)
        Assert.assertEquals(result.error, 1, errors.size.toLong())
        val ex = errors[0]
        val report = ex.scopedErrorReport
        val errorFile = requireErrorFile(report)
        Assert.assertEquals(
            File(
                projectRoot,
                "/app/src/main/res/layout/layout_with_two_way_event_attribute.xml"
            ).canonicalFile,
            errorFile.canonicalFile
        )
        Assert.assertEquals(
            String.format(ErrorMessages.TWO_WAY_EVENT_ATTRIBUTE, "android:textAttrChanged"),
            ex.bareMessage
        )
    }

    @Test
    fun testDependantDoesNotExist() {
        loadApp()
        copyTestData(
            "layout/layout_with_dependency.xml",
            "app/src/main/res/layout/layout_with_dependency.xml"
        )
        copyTestData(
            "androidx/databinding/compilationTests/badJava/ObservableNoDependent.java",
            "app/src/main/java/androidx/databinding/compilationTest/badJava/MyObservable.java"
        )
        val result: CompilationResult = assembleDebug()
        Assert.assertNotEquals(0, result.resultCode.toLong())
        val errors = ScopedException.extractErrors(result.error)
        Assert.assertEquals(result.error, 1, errors.size.toLong())
        val ex = errors[0]
        val report = ex.scopedErrorReport
        val errorFile = requireErrorFile(report)
        Assert.assertEquals(
            File(
                projectRoot,
                "/app/src/main/res/layout/layout_with_dependency.xml"
            ).canonicalFile,
            errorFile.canonicalFile
        )
        Assert.assertEquals(
            ("Could not find dependent property 'notExist' referenced in " +
                    "@Bindable annotation on " +
                    "androidx.databinding.compilationTest.badJava.MyObservable.getField"),
            ex.bareMessage
        )
    }

    @Test
    fun testDependantNotBindable() {
        loadApp()
        copyTestData(
            "layout/layout_with_dependency.xml",
            "app/src/main/res/layout/layout_with_dependency.xml"
        )
        copyTestData(
            "androidx/databinding/compilationTests/badJava/ObservableNotBindableDependent.java",
            "app/src/main/java/androidx/databinding/compilationTest/badJava/MyObservable.java"
        )
        val result: CompilationResult = assembleDebug()
        Assert.assertNotEquals(0, result.resultCode.toLong())
        val errors = ScopedException.extractErrors(result.error)
        Assert.assertEquals(result.error, 1, errors.size.toLong())
        val ex = errors[0]
        val report = ex.scopedErrorReport
        val errorFile = requireErrorFile(report)
        Assert.assertEquals(
            File(
                projectRoot,
                "/app/src/main/res/layout/layout_with_dependency.xml"
            ).canonicalFile,
            errorFile.canonicalFile
        )
        Assert.assertEquals(
            ("The dependent property 'otherField' referenced in " +
                    "@Bindable annotation on " +
                    "androidx.databinding.compilationTest.badJava.MyObservable.getField " +
                    "must be annotated with @Bindable"), ex.bareMessage
        )
    }

    @Test
    fun testDependantField() {
        loadApp()
        copyTestData(
            "layout/layout_with_dependency.xml",
            "app/src/main/res/layout/layout_with_dependency.xml"
        )
        copyTestData(
            "androidx/databinding/compilationTests/badJava/ObservableFieldDependent.java",
            "app/src/main/java/androidx/databinding/compilationTest/badJava/MyObservable.java"
        )
        val result: CompilationResult = assembleDebug()
        Assert.assertNotEquals(0, result.resultCode.toLong())
        val errors = ScopedException.extractErrors(result.error)
        Assert.assertEquals(result.error, 1, errors.size.toLong())
        val ex = errors[0]
        val report = ex.scopedErrorReport
        val errorFile = requireErrorFile(report)
        Assert.assertTrue(errorFile.exists())
        Assert.assertEquals(
            File(
                projectRoot,
                "/app/src/main/res/layout/layout_with_dependency.xml"
            ).canonicalFile,
            errorFile.canonicalFile
        )
        Assert.assertEquals(
            ("Bindable annotation with property names is only supported on methods. " +
                    "Field 'androidx.databinding.compilationTest.badJava.MyObservable.field' has " +
                    "@Bindable(\"otherField\")"), ex.bareMessage
        )
    }

    private fun assertMessages(
        exceptions: List<ScopedException>,
        vararg messages: String
    ) {
        val actual = exceptions
            .map { obj: ScopedException -> obj.bareMessage }
            .toList()
        MatcherAssert.assertThat(actual, CoreMatchers.hasItems(*messages))
        MatcherAssert.assertThat(actual.size, CoreMatchers.`is`(messages.size))
    }

    private fun singleFileWarningTest(
        resource: String,
        targetFile: String,
        expectedMessage: String
    ) {
        loadApp()
        copyTestData(resource, targetFile)
        val result = assembleDebug()
        val warnings = result.bindingWarnings
        var found = false
        for (warning in warnings) {
            found = found or warning.contains(expectedMessage)
        }
        Assert.assertTrue(Joiner.on("\n").join(warnings), found)
    }

    private fun singleFileErrorTest(
        resource: String,
        targetFile: String,
        expectedExtract: String?,
        errorMessage: String?
    ): ScopedException {
        loadApp()
        copyTestData(resource, targetFile)
        val result = assembleDebug()
        assertFalse(result.isBuildSuccessful)
        val scopedException = result.bindingException
        val report = scopedException.scopedErrorReport
        assertNotNull(report)
        assertEquals(1, report.locations.size.toLong())

        val loc = report.locations[0]
        if (expectedExtract != null) {
            val extract: String = extract(targetFile, loc)
            Assert.assertEquals(scopedException.message, expectedExtract, extract)
        }
        val errorFile: File = requireErrorFile(report)
        Assert.assertEquals(
            File(projectRoot, targetFile).canonicalFile,
            errorFile.canonicalFile
        )
        if (errorMessage != null) {
            Assert.assertEquals(errorMessage, scopedException.bareMessage)
        }
        return scopedException
    }

    /**
     * Extracts the text in the given location from the file at the given application path.
     *
     * @param relativePath the relative path of the file to be extracted from
     * @param location  The location to extract
     * @return The string that is contained in the given location
     * @throws IOException If file is invalid.
     */
    private fun extract(relativePath: String, location: Location): String {
        val file = File(projectRoot, relativePath)
        Assert.assertTrue(file.exists())
        val result = StringBuilder()
        val lines = file.readLines(StandardCharsets.UTF_8)
        for (i in location.startLine..location.endLine) {
            if (i > location.startLine) {
                result.append("\n")
            }
            val line = lines[i]
            var start = 0
            if (i == location.startLine) {
                start = location.startOffset
            }
            var end = line.length - 1 // inclusive
            if (i == location.endLine) {
                end = location.endOffset
            }
            result.append(line.substring(start, end + 1))
        }
        return result.toString()
    }

    /**
     * Finds the error file referenced in the given error report.
     * Handles possibly relative paths.
     *
     * Throws an assertion exception if the error file reported cannot be found.
     */
    private fun requireErrorFile(report: ScopedErrorReport): File {
        val path = report.filePath
        Assert.assertNotNull(path)
        var file = File(path)
        if (file.exists()) {
            return file
        }
        // might be relative, try in test project folder
        file = File(projectRoot, path)
        Assert.assertTrue("required error file is missing in " + file.absolutePath, file.exists())
        return file
    }
}
